File: src/graph/dfs/km104.py
from typing import Dict, List, Set, Tuple


rowNum, colNum = map(int, input().strip().split(" "))

step: List[Tuple[int, int]] = [(1, 0), (0, 1), (-1, 0), (0, -1)]

# visit = [[False for _ in range(colNum)] for _ in range(rowNum)]

grid: List[List[int]] = []

mark: int = 1
allIsland: bool = True

for _ in range(rowNum):
    grid.append(list(map(int, input().strip().split(" "))))

island2size: Dict[int, int] = dict()

size: int = 0


def dfs(y: int, x: int) -> None:
    global size
    size += 1
    grid[y][x] = mark

    for item in step:
        dy, dx = item[0], item[1]
        nextY, nextX = y + dy, x + dx
        if (
            nextY >= 0
            and nextY < rowNum
            and nextX >= 0
            and nextX < colNum
            and grid[nextY][nextX] == 1
        ):
            dfs(nextY, nextX)


for y in range(rowNum):
    for x in range(colNum):
        if grid[y][x] == 1:
            mark += 1
            size = 0
            dfs(y, x)
            island2size[mark] = size
        if grid[y][x] == 0:
            allIsland = False

if allIsland:
    print(rowNum * colNum)
    exit(0)

ans: int = 1

for y in range(rowNum):
    for x in range(colNum):
        if grid[y][x] == 0:
            cur = 1
            neighborMark: Set[int] = set()
            for item in step:
                dy, dx = item
                neighborY = y + dy
                neighborX = x + dx

                if (
                    neighborY >= 0
                    and neighborY < rowNum
                    and neighborX >= 0
                    and neighborX < colNum
                    and grid[neighborY][neighborX] > 1
                ):
                    neighborMark.add(grid[neighborY][neighborX])

            for mark in neighborMark:
                cur += island2size[mark]

            ans = max(ans, cur)

print(ans)

File: src/graph/dfs/km104.ts
import { createInterface } from "readline";

const rl = createInterface({
  input: process.stdin,
  output: process.stdout,
});

let lineno = 0;
let rowNum = 0;
let colNum = 0;

let mark = 2;
let size = 0;
const island2size = new Map<number, number>();

const grid: number[][] = [];

let allIsland = true;

let ans = 1;

const step: [number, number][] = [
  [0, 1],
  [1, 0],
  [0, -1],
  [-1, 0],
];

rl.on("line", (line: string) => {
  lineno++;

  if (lineno === 1) {
    [rowNum, colNum] = line
      .trim()
      .split(" ")
      .map((item) => Number.parseInt(item));
    return;
  }

  grid.push(
    line
      .trim()
      .split(" ")
      .map((item) => Number.parseInt(item)),
  );

  if (lineno === rowNum + 1) {
    rico();
    rl.close();
  }
});

function dfs(y: number, x: number) {
  size++;
  grid[y][x] = mark;

  for (const [dy, dx] of step) {
    const nextY = y + dy;
    const nextX = x + dx;

    if (
      nextY >= 0 &&
      nextY < rowNum &&
      nextX >= 0 &&
      nextX < colNum &&
      grid[nextY][nextX] === 1
    ) {
      dfs(nextY, nextX);
    }
  }
}

function rico() {
  for (let y = 0; y < rowNum; y++) {
    for (let x = 0; x < colNum; x++) {
      if (grid[y][x] === 0) {
        allIsland = false;
      }

      if (grid[y][x] === 1) {
        size = 0;
        mark++;
        dfs(y, x);
        island2size.set(mark, size);
      }
    }
  }

  if (allIsland) {
    // console.log(rowNum * colNum)
    process.stdout.write(`${rowNum * colNum}\n`);
    process.exit(0);
  }

  for (let y = 0; y < rowNum; y++) {
    for (let x = 0; x < colNum; x++) {
      if (grid[y][x] === 0) {
        const markSet = new Set<number>();

        for (const [dy, dx] of step) {
          const siblingY = y + dy;
          const siblingX = x + dx;
          if (
            siblingY >= 0 &&
            siblingY < rowNum &&
            siblingX >= 0 &&
            siblingX < colNum &&
            grid[siblingY][siblingX] > 1
          ) {
            markSet.add(grid[siblingY][siblingX]);
          }
        }

        let cur = 1;
        for (const mark of markSet) {
          cur += island2size.get(mark) ?? 0;
        }

        ans = Math.max(ans, cur);
      }
    }
  }

  process.stdout.write(`${ans}\n`);
}
